### PR TITLE
sphinx 2.2.11-1.1 debian stretch

### PR DESCRIPTION
**Pull request by ltclm**
related to https://github.com/geoadmin/service-sphinxsearch/pull/375

- [ ] ok emineo
- [ ] ok derpunkt

### CODE REVIEW

**Comment by ltclm** [comment]
@gjn @davidoesch @danduk82 
here's a first testlink with the new sphinx version 2.2.11-1.1 on debian stretch.
https://mf-chsdi3.dev.bgdi.ch/dev_ltclm_migrate_sphinx/shorten/7d3bbc51fd
sphinx changes: https://github.com/geoadmin/service-sphinxsearch/pull/375

all the chsdi integrations tests are now passing without error.
**BUT there is still a timeout on the sphinxsearch service** when trying to search in swisssearch for something unprecise like:
```
strasse 5
``` 

some sphinx related info on this query:
```
# query fired by sphinx api
[swisssearch] @detail "strasse 5" | @detail "^strasse 5" | @detail "strasse 5$" | @detail "^strasse 5$" | @detail "strasse 5"~5 | @detail "strasse* 5" | @detail "^strasse* 5" | @detail "strasse* 5"~5 | @detail "*strasse* 5" | @detail "^*strasse* 5" | @detail "*strasse* 5"~5 | @detail "strasse 5*" | @detail "^strasse 5*" | @detail "strasse* 5*" | @detail "strasse* 5*"~5 | @detail "*strasse* 5*"

# address headers sphinx on new index
$ indextool --dumpheader address
Sphinx 2.2.11-id64-release (95ae9a6)
Copyright (c) 2001-2016, Andrew Aksyonoff
Copyright (c) 2008-2016, Sphinx Technologies Inc (http://sphinxsearch.com)

using config file '/etc/sphinxsearch/sphinx.conf'...
dumping header for index 'address'...
version: 42
idbits: 64
docinfo: extern
fields: 2
  field 0: detail
  field 1: geom_quadindex
attrs: 16
  attr 0: feature_id, string, bitoff 0
  attr 1: detail, string, bitoff 32
  attr 2: label, string, bitoff 64
  attr 3: origin, string, bitoff 96
  attr 4: geom_quadindex, string, bitoff 128
  attr 5: geom_st_box2d, string, bitoff 160
  attr 6: geom_st_box2d_lv95, string, bitoff 192
  attr 7: rank, uint, bitoff 224
  attr 8: x, float, bitoff 256
  attr 9: y, float, bitoff 288
  attr 10: y_lv95, float, bitoff 320
  attr 11: x_lv95, float, bitoff 352
  attr 12: lat, float, bitoff 384
  attr 13: lon, float, bitoff 416
  attr 14: num, uint, bitoff 448
  attr 15: zoomlevel, uint, bitoff 480
total-documents: 2129696
total-bytes: 155652605
total-duplicates: 0
min-prefix-len: 0
min-infix-len: 1
max-substring-len: 0
exact-words: 0
html-strip: 0
html-index-attrs: (null)
html-remove-elements: (null)
index-zones: (null)
tokenizer-type: 2
tokenizer-case-folding: (null)
tokenizer-min-word-len: 1
tokenizer-ngram-chars: (null)
tokenizer-ngram-len: 0
tokenizer-exceptions: (null)
tokenizer-phrase-boundary: (null)
tokenizer-ignore-chars: (null)
tokenizer-blend-chars: (null)
tokenizer-blend-mode: (null)
tokenizer-exceptions: (null)
dictionary-embedded-exceptions: 1
dict: keywords
dictionary-morphology: (null)
dictionary-stopwords-file: (null)
dictionary-embedded-stopwords: 1
dictionary-wordform-file [0]: /var/lib/sphinxsearch/data/index/wordforms_main.txt
dictionary-embedded-wordforms: 1
        dictionary-embedded-wordform [0]: w > weg
        dictionary-embedded-wordform [1]: al > allee
        dictionary-embedded-wordform [2]: str > strasse
        dictionary-embedded-wordform [3]: gi > industriegebiet
        dictionary-embedded-wordform [4]: ch > chemin
        dictionary-embedded-wordform [5]: av > avenue
        dictionary-embedded-wordform [6]: boul > boulevard
        dictionary-embedded-wordform [7]: bd > boulevard
        dictionary-embedded-wordform [8]: rte > route
        dictionary-embedded-wordform [9]: faub > faubourg
        dictionary-embedded-wordform [10]: fbg > faubourg
        dictionary-embedded-wordform [11]: r > rue
        dictionary-embedded-wordform [12]: sent > sentier
        dictionary-embedded-wordform [13]: pass > passage
        dictionary-embedded-wordform [14]: v > viale
        dictionary-embedded-wordform [15]: p > piazza
        dictionary-embedded-wordform [16]: pl > pla
min-stemming-len: 1
killlist-size: 0
min-max-index: 38334528

# address headers sphinx on old index
$ indextool --dumpheader address
Sphinx 2.1.6-id64-release (rel21-r4571)
Copyright (c) 2001-2014, Andrew Aksyonoff
Copyright (c) 2008-2014, Sphinx Technologies Inc (http://sphinxsearch.com)

using config file '/etc/sphinxsearch/sphinx.conf'...
dumping header for index 'address'...
version: 38
idbits: 64
docinfo: extern
fields: 2
  field 0: detail
  field 1: geom_quadindex
attrs: 16
  attr 0: feature_id, string, bitoff 0
  attr 1: detail, string, bitoff 32
  attr 2: label, string, bitoff 64
  attr 3: origin, string, bitoff 96
  attr 4: geom_quadindex, string, bitoff 128
  attr 5: geom_st_box2d, string, bitoff 160
  attr 6: geom_st_box2d_lv95, string, bitoff 192
  attr 7: rank, uint, bitoff 224
  attr 8: x, float, bitoff 256
  attr 9: y, float, bitoff 288
  attr 10: y_lv95, float, bitoff 320
  attr 11: x_lv95, float, bitoff 352
  attr 12: lat, float, bitoff 384
  attr 13: lon, float, bitoff 416
  attr 14: num, uint, bitoff 448
  attr 15: zoomlevel, uint, bitoff 480
total-documents: 2129696
total-bytes: 155652605
min-prefix-len: 0
min-infix-len: 1
max-substring-len: 0
exact-words: 0
html-strip: 0
html-index-attrs: (null)
html-remove-elements: (null)
index-zones: (null)
tokenizer-type: 2
tokenizer-case-folding: (null)
tokenizer-min-word-len: 1
tokenizer-ngram-chars: (null)
tokenizer-ngram-len: 0
tokenizer-exceptions: (null)
tokenizer-phrase-boundary: (null)
tokenizer-ignore-chars: (null)
tokenizer-blend-chars: (null)
tokenizer-blend-mode: (null)
dictionary-morphology: (null)
dictionary-stopwords: (null)
        dictionary-wordforms [0]: /var/lib/sphinxsearch/data/index/wordforms_main.txt
min-stemming-len: 1
killlist-size: 0
min-max-index: 38334528

```


Any help, hint is appreciated

**Comment by ltclm** [comment]
@gjn @davidoesch @danduk82
the testlink with the new sphinx version 2.2.11-1.1 on debian stretch is now working, all the tests are passing and the timeout issues have been solved.
https://mf-chsdi3.dev.bgdi.ch/dev_ltclm_migrate_sphinx/shorten/7d3bbc51fd
sphinx changes: geoadmin/service-sphinxsearch#375

@davidoesch @gjn maybe this should be tested by some clients too.

**Comment by davidoesch** [comment]
Needs to be tested by e Umzug. Can i send them the testlink above?

**Comment by ltclm** [comment]
@davidoesch yes, but i think we should test it internally first.

**Comment by davidoesch** [comment]
Ok .. we need to brainstorm a testbook . Will brain storm with you the current usage. Is it possible to extract from current prod via kibana the last couple 1000000 swissearch requests and parse it through the new Sphinx search and check if all will pass?

**Comment by ltclm** [comment]
sounds like an interesting athena excercise ;-)
yes we should do that, i will prepare the list of queries.

**Comment by ltclm** [comment]
i have extracted some searchserver queries from the logs of 2018-09-27 - 2018-09-30.
The file has `1181019 requests ` and can be downloaded from 
`s3://swisstopo-external-filesharing/searchserver_samples.txt`


There are only succesful requests (status 200) from production logs.

**Comment by davidoesch** [comment]
@gjn : IMHO if 1181019 requests are successfull : lgtm

**Comment by ltclm** [comment]
tests are not yet done, i will post the results here.

**Comment by ltclm** [comment]
@davidoesch @danduk82 @gjn 

I did a test with the same 100'000 requests on mf-chsdi3.dev on the old and the new sphinx version / code. The test run with 1'000'000 requests takes too long and the test environment is not stable enough (apache restarts, db deploys, sphinx index, network...)

Here the results for 100'000 requests:
### sphinx 2.2.11 debian stretch
https://mf-chsdi3.dev.bgdi.ch/dev_ltclm_migrate_sphinx/
Total time for 100'000 requests: `7h28'` 
http status stats:
```
  98624 200
    130 404
     91 500
   1155 504
```

### sphinx 2.1.6 debian wheezy
https://mf-chsdi3.dev.bgdi.ch/
Total time for 100'000 requests: `7h23'` 
http status stats:
```
  99967 200
     32 500
      1 504
```

The quality of the new service is not good enough, we can either
* try to fix the remaining issues with the new service (timeouts `504` error `404`)
* find a backport package on stretch providing the same sphinx version (2.1.6)
* other suggestions

**Comment by ltclm** [comment]
apparently remy has already done a sphinx migration and has found a workaround for the timeout issues with dict=keywords
http://sphinxsearch.com/forum/view.html?id=13960

Im trying to go into that direction...

**Comment by ltclm** [comment]
* All the swisssearch index have been changed to type `dict=crc` https://github.com/geoadmin/service-sphinxsearch/pull/375/commits/e77720a028ab201e743de8b05d75fa36739c3b31
* The chsdi sphinx client code has been reverted to the old version

the service is much more stable now and there are almost no timeouts anymore.
Im currently re-running the test script.

**Comment by ltclm** [comment]
the results of the new dict=crc swisssearch index is promising:
Total time for 100'000 requests: `7h23'`
http status stats:
```
  99938 200
     62 504
```
I think we're ready to go.
@davidoesch @danduk82 @gjn agree?

If yes we can ask e Umzug for some tests.

**Comment by danduk82** [comment]
Seems good to me. Thanks a lot for the effort @ltclm 

@davidoesch , can you ask Umzug to test?
test link: https://mf-chsdi3.dev.bgdi.ch/dev_ltclm_migrate_sphinx/

File: rc_dev [hunk 1]
```diff
@@ -17,6 +17,7 @@ export PUBLIC_BUCKET_HOST=public.dev.bgdi.ch
 export ROBOTS_FILE=robots.txt
 export SERVER_PORT=6543
 export SPHINXHOST=service-sphinxsearch.dev.bgdi.ch
+export SPHINXHOST=10.220.4.170
```
Comment (danduk82) [inline rc_dev]: please do not forget to let C2C create a new instance for dev with correct puppet resources and test that the install is perfectly functional before creating the new int and prod machines.

**Comment by davidoesch** [comment]
Will ask e umzug on Monday. Don't push it live until we got a feedback from them

**Comment by gjn** [comment]
We have quantitative analysis.

Did we do a qualitative analysis? The search was quite fine tuned in the past we some special cases, mainly regarding result matching and sorting. I want this tested and documented before we go live with this. (We need to be able to explain differences, if there are any)

**Comment by ltclm** [comment]
@gjn the order or the results is changing.
for swisseach (exact match) we are using:
* ranking mode: SPH_RANK_WORDCOUNT -> used for calulating relevance aka weight
* sort mode: SPH_SORT_EXTENDE `rank ASC, @weight DESC, num ASC`
rank attribute to group results/priority (gazetteer->address->...)
weight calculated by sphinx ranker
num numeric part of addresses (streetnumber)

the default order of results within the same rank, weight, num interval has been changed between sphinx 2.1.6 and 2.2.11

If you compare the first three results of:

http://mf-chsdi3.dev.bgdi.ch/rest/services/ech/SearchServer?sr=2056&searchText=avenue+2&lang=de&type=locations

http://mf-chsdi3.dev.bgdi.ch/dev_ltclm_migrate_sphinx/rest/services/ech/SearchServer?sr=2056&searchText=avenue+2&lang=de&type=locations

you will see that the order of this 3 objects with same weight, rank and num value has changed.
**As far as i know there is no way to change this default order.**

the sphinx search query sent to the service by chsdi is:
```
@detail "avenue 2" | @detail "^avenue 2" | @detail "avenue 2$" | @detail "^avenue 2$" | @detail "avenue 2"~5 | @detail "avenue* 2" | @detail "^avenue* 2" | @detail "avenue* 2"~5 | @detail "*avenue* 2" | @detail "^*avenue* 2" | @detail "*avenue* 2"~5 | @detail "avenue 2*" | @detail "^avenue 2*" | @detail "avenue* 2*" | @detail "avenue* 2*"~5 | @detail "*avenue* 2*"
```

**Comment by ltclm** [comment]
#### good news:
the results without SPH_RANKER_WORDCOUNT between 2.1.6 and 2.2.11 are 100% identical.

#### bad news:
i did some more tests with sphinx docker images of the versions 2.3.2 and 3.0.3
The internal configuration of SPH_RANKER_WORDCOUNT did change somewhere between the versions 2.1.6 and 2.2.11 (hard to say which change introduced the new behaviour).
Until about version 2.1.6  the following (see docs) was correct 
```
SPH_RANK_WORDCOUNT = sum(hit_count*user_weight)
```
**Thats not the case anymore.**
This change has been probably introduced with 2.2.1 and there's an open issue:
http://sphinxsearch.com/bugs/view.php?id=1733
It has been fixed in Version 3.0.3 but migrating from 2.x to 3.0.3 is a LOT of work (sphinx config has to be rebuild from scratch), and im not sure if we can rebuild the same behaviour.

**Comment by ltclm** [comment]
@davidoesch @danduk82 
any news from the customer?
Infra and scripts have been prepared to finish this migration...

**Comment by davidoesch** [comment]
no feedback from customer

-> A go for you @ltclm

**Pull request merged**